common/src/main/java/com/microsoft/identity/common/internal/ui/webview/AzureActiveDirectoryWebViewClient.java

```diff
@@ -214,35 +214,40 @@ protected boolean processRedirectUrl(@NonNull final WebView view, @NonNull final
-            // Also it must handle 3rd party MDMs as appropriate, which will depend on the browser flow determining the authority.
-            Logger.info(TAG + methodName, "It is a device CA request on IPPhone. Company Portal is installed.");
-            try {
-                Logger.verbose(TAG + methodName, "Sending intent to launch the CompanyPortal.");
-                final Intent intent = new Intent();
-                intent.setComponent(new ComponentName(AuthenticationConstants.Broker.COMPANY_PORTAL_APP_PACKAGE_NAME,
-                        AuthenticationConstants.Broker.COMPANY_PORTAL_APP_LAUNCH_ACTIVITY_NAME));
-                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
-                getActivity().startActivity(intent);
-            } catch (final SecurityException ex) {
-                Logger.warn(TAG + methodName, "Failed to launch Company Portal, falling back to browser.");
-                openLinkInBrowser(url);
+
+        view.stopLoading();
+
+        if (url.contains(AuthenticationConstants.Broker.BROWSER_DEVICE_CA_URL_QUERY_STRING_PARAMETER)) {
+            Logger.info(TAG + methodName, "This is a device CA request.");
+            final PackageHelper packageHelper = new PackageHelper(getActivity().getPackageManager());
+            final Context applicationContext = getActivity().getApplicationContext();
+
+            // If CP is installed, redirect to CP.
+            if (!getActivity().getPackageName().equalsIgnoreCase(AuthenticationConstants.Broker.COMPANY_PORTAL_APP_PACKAGE_NAME) &&
+                    packageHelper.isPackageInstalledAndEnabled(applicationContext, AuthenticationConstants.Broker.COMPANY_PORTAL_APP_PACKAGE_NAME)) {
+                try {
+                    Logger.verbose(TAG + methodName, "Sending intent to launch the CompanyPortal.");
+                    final Intent intent = new Intent();
+                    intent.setComponent(new ComponentName(AuthenticationConstants.Broker.COMPANY_PORTAL_APP_PACKAGE_NAME,
+                            AuthenticationConstants.Broker.COMPANY_PORTAL_APP_LAUNCH_ACTIVITY_NAME));
+                    intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
+                    getActivity().startActivity(intent);
+
+                    getCompletionCallback().onChallengeResponseReceived(AuthenticationConstants.UIResponse.BROWSER_CODE_MDM, new Intent());
+                    return true;
+                } catch (final Exception ex) {
+                    Logger.warn(TAG + methodName, "Failed to launch Company Portal, falling back to browser.");
+                }
             }
-        } else {
-            openLinkInBrowser(url);
+
+            // Otherwise, strip out BROWSER_DEVICE_CA_URL_QUERY_STRING_PARAMETER and launch in Browser.
+            openLinkInBrowser(url.replace(AuthenticationConstants.Broker.BROWSER_DEVICE_CA_URL_QUERY_STRING_PARAMETER, ""));
```

Comment:
Why do we need this. Shouldn't the server already be rejecting it?

Reply:
This is what I and Carlos originally agreed on. Although I agree with you that if we make the query string parameter unique enough, we might not have to do this.

I've forwarded him this PR and will let him chime in.

Reply:
Yes I would be curious to understand why server can't reject it .

Reply:
Intune confirmed they'll reject it. I'll remove this logic.